pydantic/schema.py

```diff
@@ -0,0 +1,392 @@
+    assert issubclass(model, main.BaseModel)
+    flat_models.add(model)
+    for field in model.__fields__.values():
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_field(field: Field) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    if field.sub_fields:
+        flat_models = flat_models | get_flat_models_from_sub_fields(field.sub_fields)
+    elif isinstance(field.type_, type) and issubclass(field.type_, main.BaseModel):
+        flat_models = flat_models | get_flat_models_from_model(field.type_)
+    return flat_models
+
+
+def get_flat_models_from_sub_fields(fields) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for field in fields:
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_models(models: Sequence[Type['main.BaseModel']]) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for model in models:
+        flat_models = flat_models | get_flat_models_from_model(model)
+    return flat_models
+
+
+def get_long_model_name(model: Type['main.BaseModel']):
+    assert issubclass(model, main.BaseModel)
+    prefix = model.__module__.replace('.', '__')
+    name = model.__name__
+    return f'{prefix}__{name}'
+
+
+def get_model_name_maps(
+    unique_models: Set[Type['main.BaseModel']]
+) -> Tuple[Dict[str, Type['main.BaseModel']], Dict[Type['main.BaseModel'], str]]:
+    name_model_map: Dict[str, Type[main.BaseModel]] = {}
+    conflicting_names: Set[str] = set()
+    for model in unique_models:
+        model_name = model.__name__
+        if model_name in conflicting_names:
+            model_name = get_long_model_name(model)
+            name_model_map[model_name] = model
+        elif model_name in name_model_map:
+            conflicting_names.add(model_name)
+            conflicting_model = name_model_map[model_name]
+            del name_model_map[model_name]
+            conflicting_model_name = get_long_model_name(conflicting_model)
+            name_model_map[conflicting_model_name] = conflicting_model
+            model_name = get_long_model_name(model)
+            name_model_map[model_name] = model
+        else:
+            name_model_map[model_name] = model
+    model_name_map = {v: k for k, v in name_model_map.items()}
+    return name_model_map, model_name_map
+
+
+def field_schema(
+    field: Field, *, by_alias=True, model_name_map: Dict[Type['main.BaseModel'], str], ref_prefix='#/definitions/'
+) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+    schema_overrides = False
+    s = dict(title=field._schema.title or field.alias.title())
+    if field._schema.title:
+        schema_overrides = True
+
+    if not field.required and field.default is not None:
+        schema_overrides = True
+        if isinstance(field.default, (int, float, bool, str)):
```

Comment:
no need for `bool` here, bool is a subclass of int.

Reply:
Done